### PR TITLE
Support for null strings in Collate method

### PR DESCRIPTION
**Pull request by APalpattuwar7**

### CODE REVIEW

File: Source/LinqToDB/Sql/Sql.Collate.cs [hunk 1]
```diff
@@ -51,6 +53,7 @@ internal class PostgreSQLCollationBuilder : IExtensionCallBuilder
 			public void Build(ISqExtensionBuilder builder)
 			{
 				var expr      = builder.GetExpression("expr");
+				SqlExpression.CanBeNull = expr.CanBeNull;
```
Comment (MaceWindu) [inline Source/LinqToDB/Sql/Sql.Collate.cs:56]: `expr.CanBeNull;` should be assigned to `SqlExpression` createdbelow using `new SqlExpression(typeof(string), $"{{0}} COLLATE \"{collation}\"", Precedence.Primary, expr);`

Comment (APalpattuwar7) [reply Source/LinqToDB/Sql/Sql.Collate.cs:56]: Should it be like this? - 
`SqlExpression.CanBeNull = new SqlExpression(typeof(string), $"{{0}} COLLATE \"{collation}\"", Precedence.Primary, expr);`

File: Source/LinqToDB/Sql/Sql.Collate.cs [hunk 2]
```diff
@@ -17,7 +17,8 @@ public static partial class Sql
 		[Extension("", ServerSideOnly = true, BuilderType = typeof(DB2LUWCollationBuilder)    , Configuration = ProviderName.DB2LUW)]
 		[Extension("", ServerSideOnly = true, BuilderType = typeof(PostgreSQLCollationBuilder), Configuration = ProviderName.PostgreSQL)]
 		[Extension("", ServerSideOnly = true, BuilderType = typeof(NamedCollationBuilder))]
-		public static string Collate(this string expr, [SqlQueryDependent] string collation)
+		[return: NotNullIfNotNull("expr")]
```
Comment (APalpattuwar7) [inline Source/LinqToDB/Sql/Sql.Collate.cs:20]: Which reference should I add for `NotNullIfNotNull`?
Sorry I am unable to build solution so couldn't figure it out.

Comment (MaceWindu) [reply Source/LinqToDB/Sql/Sql.Collate.cs:20]: System.Diagnostics.CodeAnalysis

File: Source/LinqToDB/Sql/Sql.Collate.cs (hunk 1 above)
Comment (MaceWindu) [reply Source/LinqToDB/Sql/Sql.Collate.cs:56]:
```cs
builder.ResultExpression = new SqlExpression(typeof(string), $"{{0}} COLLATE \"{collation}\"", Precedence.Primary, expr)
{
    CanBeNull = expr.CanBeNull
};
```

Comment (APalpattuwar7) [reply Source/LinqToDB/Sql/Sql.Collate.cs:60]: @MaceWindu Could you please review it once?

File: Source/LinqToDB/Sql/Sql.Collate.cs [hunk 3]
```diff
@@ -66,6 +74,10 @@ public void Build(ISqExtensionBuilder builder)
 
 				// collation cannot be parameter
 				builder.ResultExpression = new SqlExpression(typeof(string), $"COLLATION_KEY_BIT({{0}}, {{1}})", Precedence.Primary, expr, new SqlValue(typeof(string), collation));
+				builder.ResultExpression = new SqlExpression(typeof(string), $"{{0}} COLLATE \"{collation}\"", Precedence.Primary, expr)
```
Comment (MaceWindu) [inline Source/LinqToDB/Sql/Sql.Collate.cs:77]: that line is not correct

Comment (APalpattuwar7) [reply Source/LinqToDB/Sql/Sql.Collate.cs:77]: Shall I remove it completely?

Comment (MaceWindu) [reply Source/LinqToDB/Sql/Sql.Collate.cs:77]: your change doesn't match old value. old code generated COLLATION_KEY_BIT call, new one generates COLLATE clause

**Comment by MaceWindu** [comment]
superseeded by #3853

**Pull request closed without merging**